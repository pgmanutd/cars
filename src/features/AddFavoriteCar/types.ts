import { Car } from 'features/CarInfo/types';

export interface FavoriteCars {
  [stockNumber: string]: Car;
}
