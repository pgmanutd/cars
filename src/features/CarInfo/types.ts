export type StockNumberParam = string;

export interface Car {
  stockNumber: number;
  manufacturerName: string;
  modelName: string;
  mileage: {
    number: number;
    unit: string;
  };
  fuelType: string;
  color: string;
  pictureUrl: string;
}

export interface CarResponse {
  car: Car;
}

export interface FavoriteCars {
  [stockNumber: string]: Car;
}
