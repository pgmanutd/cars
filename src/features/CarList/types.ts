import { Car } from 'features/CarInfo/types';

export type PageQuery = string;

export type Cars = Car[];

export interface CarListResponse {
  cars: Cars;
  totalPageCount: number;
  totalCarsCount: number;
}
