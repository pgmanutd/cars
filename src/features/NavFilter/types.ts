export type ColorQuery = string;
export type ManufacturerQuery = string;

export type Color = string;
export type Colors = Color[];

export interface ColorsResponse {
  colors: Colors;
}

export interface Manufacturer {
  name: string;
  models: {
    name: string;
  }[];
}
export type Manufacturers = Manufacturer[];

export interface ManufacturersResponse {
  manufacturers: Manufacturers;
}
