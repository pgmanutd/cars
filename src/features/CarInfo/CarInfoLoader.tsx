import React from 'react';
import Grid from '@material-ui/core/Grid';
import Skeleton from '@material-ui/lab/Skeleton';

const CarInfoLoader: React.FC = () => (
  <Grid data-testid="CarInfoLoader" component="section" container spacing={4}>
    <Grid item xs={12}>
      <Skeleton
        data-testid="CarInfoImageSkeleton"
        variant="rect"
        height={200}
      />
    </Grid>
    <Grid item xs={7}>
      <Skeleton height={100} width="70%" />
      <Skeleton height={50} />
      <Skeleton variant="rect" height={100} />
    </Grid>
    <Grid item xs={5}>
      <Skeleton height={250} />
    </Grid>
  </Grid>
);

export default CarInfoLoader;
