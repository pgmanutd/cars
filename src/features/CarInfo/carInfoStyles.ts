import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    carImageContainer: {
      display: 'flex',
      justifyContent: 'center',
    },
    carImage: {
      height: 250,
      width: 300,
    },
    favoriteBox: {
      border: `1px solid ${theme.palette.divider}`,
      padding: theme.spacing(3),
    },
    favoriteButtonContainer: {
      display: 'flex',
      justifyContent: 'flex-end',
      marginTop: theme.spacing(2),
    },
  }),
);
