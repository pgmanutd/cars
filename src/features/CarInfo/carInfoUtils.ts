import { Car, CarResponse, FavoriteCars } from './types';

export const getCarInfo = ({
  isLoading,
  data,
  favoriteCars,
}: {
  isLoading: boolean;
  data: CarResponse | null;
  favoriteCars: FavoriteCars;
}) => {
  const isCarInfoLoading = isLoading || !data;
  const car = data?.car ?? ({} as Car);
  const isCarSetAsFavorite = favoriteCars[car.stockNumber];

  return {
    isCarInfoLoading,
    car,
    isCarSetAsFavorite,
  };
};
