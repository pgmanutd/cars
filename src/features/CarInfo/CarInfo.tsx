import React, { memo } from 'react';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import Grid from '@material-ui/core/Grid';
import CardMedia from '@material-ui/core/CardMedia';
import { Redirect } from 'react-router-dom';

import apiPaths from 'constants/apiPaths';
import routePaths from 'constants/routePaths';

import { getCarName, getCarFeatures } from 'utils/carUtils';

import useFetch from 'hooks/useFetch';

import { useTranslate } from 'features/Translate';
import ToggleFavoriteCar from 'features/ToggleFavoriteCar';

import { StockNumberParam, CarResponse } from './types';
import CarInfoLoader from './CarInfoLoader';
import { getCarInfo } from './carInfoUtils';
import { useStyles } from './carInfoStyles';

export interface CarInfoProps {
  stockNumber: StockNumberParam;
}

const CarInfo: React.FC<CarInfoProps> = ({ stockNumber, ...restProps }) => {
  const classes = useStyles();
  const { translate } = useTranslate();
  const [{ isLoading, data, error }] = useFetch<CarResponse>({
    url: apiPaths.carDetails({ stockNumber }),
  });

  const { isCarInfoLoading, car } = getCarInfo({
    isLoading,
    data,
  });

  if (error) {
    return <Redirect to={routePaths.notFound} />;
  }

  if (isCarInfoLoading) {
    return <CarInfoLoader />;
  }

  return (
    <Grid
      data-testid="CarInfo"
      {...restProps}
      component="section"
      container
      spacing={4}
    >
      <Grid item md={12} xs={12}>
        <Box className={classes.carImageContainer}>
          {car.pictureUrl ? (
            <CardMedia
              image={car.pictureUrl}
              title={car.modelName}
              className={classes.carImage}
            />
          ) : (
            car.modelName
          )}
        </Box>
      </Grid>
      <Grid item md={7} xs={12}>
        <Box data-testid="CarInfoFeatures">
          <Typography variant="h3" gutterBottom>
            {getCarName(car)}
          </Typography>
          <Typography variant="body1" gutterBottom>
            {getCarFeatures(car, {
              stockNumberPrefix: translate(
                'features.CarInfo.stockNumberPrefix',
              ),
            })}
          </Typography>
          <Typography variant="body2">
            {translate('features.CarInfo.description')}
          </Typography>
        </Box>
      </Grid>
      <Grid item md={5} xs={12}>
        <ToggleFavoriteCar car={car} />
      </Grid>
    </Grid>
  );
};

export default memo(CarInfo);
