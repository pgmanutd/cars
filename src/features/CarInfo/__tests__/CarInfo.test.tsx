import React from 'react';
import { waitForElement } from '@testing-library/react';
import { Route } from 'react-router-dom';

import { renderWithProviders } from 'utils/testUtils';

import apiPaths from 'constants/apiPaths';
import routePaths from 'constants/routePaths';

import { StockNumberParam } from '../types';
import CarInfo from '../CarInfo';

describe('<CarInfo />', () => {
  const setup = ({ stockNumber }: { stockNumber: StockNumberParam }) => {
    const initialRoute = `/details/${stockNumber}`;

    const { renderResult, history } = renderWithProviders(
      <>
        <Route path={routePaths.notFound}>
          <div data-testid="NotFound">Not Found</div>
        </Route>
        <Route path={initialRoute}>
          <CarInfo stockNumber={stockNumber} />
        </Route>
      </>,
      {
        routerConfig: {
          route: initialRoute,
        },
      },
    );

    return {
      renderResult,
      history,
    };
  };

  beforeEach(() => {
    fetchMock.resetMocks();
  });

  it('should render the component and matches it against stored snapshot for success response', async () => {
    const stockNumber = '1234';
    const successResponse = {
      car: {
        stockNumber: 78924,
        manufacturerName: 'Fiat',
        modelName: 'Sedici',
        color: 'white',
        mileage: { number: 141474, unit: 'km' },
        fuelType: 'Diesel',
        pictureUrl: 'https://dummyimage.com/300/09f/fff.png',
      },
    };

    fetchMock
      .doMockOnceIf(apiPaths.carDetails({ stockNumber }))
      .mockResponseOnce(JSON.stringify(successResponse));

    const { renderResult } = setup({ stockNumber });

    await waitForElement(() => renderResult.queryByTestId('CarInfoFeatures'));

    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should render the component and matches it against stored snapshot for blank response', async () => {
    const stockNumber = '1234';
    const successResponse = null;

    fetchMock
      .doMockOnceIf(apiPaths.carDetails({ stockNumber }))
      .mockResponseOnce(JSON.stringify(successResponse));

    const { renderResult } = setup({ stockNumber });

    await waitForElement(() =>
      renderResult.queryByTestId('CarInfoImageSkeleton'),
    );

    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should redirect to not found page for failed response', async () => {
    const stockNumber = '1234';
    const errorResponse = new Error('Some error');

    fetchMock
      .doMockOnceIf(apiPaths.carDetails({ stockNumber }))
      .mockRejectOnce(errorResponse);

    const { renderResult, history } = setup({ stockNumber });

    await waitForElement(() => renderResult.queryByTestId('NotFound'));

    expect(history.location.pathname).toBe(routePaths.notFound);
  });
});
