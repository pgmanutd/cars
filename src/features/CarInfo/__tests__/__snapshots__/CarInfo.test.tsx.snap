// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarInfo /> should render the component and matches it against stored snapshot for blank response 1`] = `
<DocumentFragment>
  <section
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    data-testid="CarInfo"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
        data-testid="CarInfoImageSkeleton"
        style="height: 200px;"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
        style="width: 50%; height: 50px;"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
        style="height: 50px;"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
        style="height: 150px;"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
        style="height: 250px;"
      />
    </div>
  </section>
</DocumentFragment>
`;

exports[`<CarInfo /> should render the component and matches it against stored snapshot for success response 1`] = `
<DocumentFragment>
  <section
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    data-testid="CarInfo"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiBox-root MuiBox-root-117 makeStyles-carImageContainer-1"
      >
        <div
          class="MuiCardMedia-root makeStyles-carImage-2"
          style="background-image: url(https://dummyimage.com/300/09f/fff.png);"
          title="Sedici"
        />
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
    >
      <div
        class="MuiBox-root MuiBox-root-121"
        data-testid="CarInfoFeatures"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom"
        >
          Fiat Sedici
        </h3>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
        >
          Stock #78924 - 141474 KM - Diesel - White
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          This car is currently available and can be delivered as soon as tomorrow morning. Please be aware that delivery times shown in this page are not definitive and may change due to bad weather conditions.
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
    >
      <div
        class="MuiBox-root MuiBox-root-152 makeStyles-favoriteBox-3"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          If you like this car, click the button and save it in your collection of favorite items.
        </p>
        <div
          class="MuiBox-root MuiBox-root-153 makeStyles-favoriteButtonContainer-4"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Save
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
